### PR TITLE
[TIR] Unify index data type when creating prim func

### PR DESCRIPTION
**Pull request by vinx13**
* Added data type pass unification pass to by default promote data types of all indices and shapes to int64 when creating prim func.
* Added some fixes for lowering passes to make it compatible with int64 data type.

### CODE REVIEW

**Comment by junrushao** [comment]
Happy to review and let's fix the CI :-)

File: src/tir/ir/data_type_rewriter.cc [hunk 1]
```diff
@@ -191,5 +191,352 @@ PrimExpr DataTypeLegalizer::VisitExpr_(const CallNode* op) {
...
+#define DEFINE_CMPOP_EXPR_MUTATE_WITH_TYPE_MATCH(OP, FUNC)                         \
+  PrimExpr IndexDataTypeRewriter::VisitExpr_(const OP* op) {                       \
+    bool is_enabled = is_enabled_;                                                 \
+    is_enabled_ = is_condition_ && op->a->dtype.is_int() && op->b->dtype.is_int(); \
+    auto result = Parent::VisitExpr_(op);                                          \
+    is_enabled_ = is_enabled;                                                      \
+    return std::move(result);                                                      \
+  }
+
+DEFINE_CMPOP_EXPR_MUTATE_WITH_TYPE_MATCH(EQNode, operator==);
```
Comment (junrushao) [inline src/tir/ir/data_type_rewriter.cc:487]: Prefix it with TVM_

File: src/tir/transforms/narrow_datatype.cc [hunk 2]
```diff
@@ -315,65 +265,25 @@ class DataTypeRewriter : public DataTypeLegalizer {
...
-Stmt NarrowDataType(Stmt stmt, int target_bits) { return DataTypeRewriter(target_bits)(stmt); }
-
 namespace transform {
 
 Pass NarrowDataType(int target_bits) {
   auto pass_func = [target_bits](PrimFunc f, IRModule m, PassContext ctx) {
     auto* n = f.CopyOnWrite();
-    n->body = DataTypeRewriter(target_bits)(std::move(n->body));
+    n->body = NarrowDataTypeRewriter(target_bits)(std::move(n->body));
+    // LOG(INFO) << "AfterNarrow: " << tir::AsTVMScript(f);
```
Comment (junrushao) [inline src/tir/transforms/narrow_datatype.cc:286]: Remove this line?

**Review by junrushao — APPROVED** [review]
LGTM!

**Comment by vinx13** [comment]
Sometimes the model contains mixed indices type (e.g both int32 and int64). It causes dtype mismatch error during scheduling. It is expected that this pass doesn't hurt performance since there is another pass NarrowDataType that should convert it back to int32 (or other smaller types) if possible.

**Pull request merged**